package com.adonai.wallet.sync;

message SyncRequest {
    required string account = 1;
    required string password = 2;
    
    enum SyncType {
        REGISTER = 0; // server will listen to register request, to auth request otherwise
        MERGE = 1;
    }
    
    optional SyncType syncType = 3 [default = MERGE];
}

message SyncResponse {
    enum SyncAck {
        OK = 200;
        ACCOUNT_EXISTS = 402;
        AUTH_WRONG = 403;
        UNKNOWN_ERROR = 999;
    }
    
    required SyncAck syncAck = 1;
}

message AccountRequest {
    repeated int64 knownID = 1; // known to client server IDs to detect deletions
}

message AccountResponse {
    repeated int64 deletedID = 1; // accounts deleted on server/client
    repeated Account account = 2; // accounts added on server/client
}

message AccountAck {
    repeated int64 deletedGuid = 1; // accounts deleted on server as confirmation
    repeated int64 writtenGuid = 2; // accounts written on server as confirmation
}

message Account {
    required uint64 ID = 1;
    required string name = 2;
    optional string description = 3;
    required string currency = 4;
    required string amount = 5; // BigDecimal for Java
    optional int32 color = 6;
}